### PR TITLE
[Fast Refresh] Support injecting runtime after renderer executes

### PR DESCRIPTION
**Pull request by gaearon**
I think this fixes https://github.com/facebook/react/issues/17626 and https://github.com/facebook/react/issues/17552.

I was assuming that Fresh runtime always executes before the renderer. But on the web this won't be the case if you load React from CDN and *then* run your webpack bundle (https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/13). And separating `<script>` tags can be annoying and complicated.

Instead, we can handle the simple case — when ReactDOM executes before the webpack bundle — gracefully. The DevTools global hook already has a [renderers map](https://github.com/facebook/react/blob/36a6e29bb3eead85e3500ba7269cbcd55516a8fb/packages/react-devtools-shared/src/hook.js#L280) so all we need to connect to them is to read from it. In fact that's how DevTools backends themselves [connect to already injected renderers](https://github.com/facebook/react/blob/36a6e29bb3eead85e3500ba7269cbcd55516a8fb/packages/react-devtools-shared/src/backend/index.js#L85-L86).

This still doesn't handle the case where you mount some roots *before* injecting the runtime. In that case they won't be "visible" to Fresh. Arguably this is more of a corner case and I wouldn't bother solving this yet. The webpack plugin should be able to ensure that injection happens before the actual app code runs.

I added a regression test.

### CODE REVIEW

File: packages/react-refresh/src/ReactFreshRuntime.js [hunk 1]
```diff
@@ -435,6 +435,7 @@ export function injectIntoGlobalHook(globalObject: any): void {
       // Otherwise, the renderer will think that there is no global hook, and won't do the injection.
       let nextID = 0;
       globalObject.__REACT_DEVTOOLS_GLOBAL_HOOK__ = hook = {
+        renderers: new Map(),
```
Comment (gaearon) [inline packages/react-refresh/src/ReactFreshRuntime.js]: This is just so the lines below don't crash when we take this code branch. This particular Map isn't actually used further.

File: packages/react-refresh/src/ReactFreshRuntime.js [hunk 2]
```diff
@@ -468,6 +469,19 @@ export function injectIntoGlobalHook(globalObject: any): void {
       return id;
     };
 
+    // Do the same for any already injected roots.
```
Comment (gaearon) [inline packages/react-refresh/src/ReactFreshRuntime.js]: Note this is almost copy paste from the code above (462–468). I could extract a function but meh.

File: packages/react-refresh/src/ReactFreshRuntime.js [hunk 3]
```diff
@@ -468,6 +469,19 @@ export function injectIntoGlobalHook(globalObject: any): void {
...
+    // This is useful if ReactDOM has already been initialized.
+    // https://github.com/facebook/react/issues/17626
+    hook.renderers.forEach((injected, id) => {
+      if (
+        typeof injected.scheduleRefresh === 'function' &&
+        typeof injected.setRefreshHandler === 'function'
+      ) {
```
Comment (mxmul) [inline packages/react-refresh/src/ReactFreshRuntime.js]: I've applied these changes locally in my application (I had a pretty identical local branch going 😅), and it doesn't seem to work. The renderer instance has none of these properties at this point (react @ 16.8.6):

Screen Shot 2019-12-16 at 5 36 32 PM

Comment (gaearon) [reply packages/react-refresh/src/ReactFreshRuntime.js]: Can you make a minimal reproducing project?

Comment (gaearon) [reply packages/react-refresh/src/ReactFreshRuntime.js]: @mxmul Oh wait, this is expected — these were only added in 16.10 or so. You need to try with the latest version.

**Comment by leidegre** [comment]
I can't get it to work unless I enable React Development Tools, is this expected behavior? Not a big deal but I can't seem to get it to do it's magic unless React Development Tools is runnig.

**Comment by gaearon** [comment]
Sorry, not enough details. Please file a new issue.
